examples/components/BANK_MARKETING_VERTICAL/traininsilo/host.py

```diff
@@ -0,0 +1,551 @@
+                    )
+                    train_metrics.step()
+
+                    if (i + 1) % num_of_batches_before_logging == 0 or (i + 1) == len(
+                        self.train_loader_
+                    ):
+                        log_message = [
+                            f"Epoch: {epoch}/{self._epochs}",
+                            f"Iteration: {i+1}/{len(self.train_loader_)}",
+                        ]
+
+                        for name, value in train_metrics.get_step().items():
+                            log_message.append(f"{name}: {value}")
+                            self.log_metrics(
+                                mlflow_client,
+                                root_run_id,
+                                name,
+                                value,
+                            )
+                        logger.info(", ".join(log_message))
+                        train_metrics.reset_step()
+
+                test_metrics = self.test()
+                log_message = [
+                    f"Epoch: {epoch}/{self._epochs}",
+                ]
+
+                # log test metrics after each epoch
+                for name, value in train_metrics.get_global().items():
+                    log_message.append(f"{name}: {value}")
+                    self.log_metrics(
+                        mlflow_client,
+                        root_run_id,
+                        name,
+                        value,
+                    )
+                for name, value in test_metrics.get_global().items():
+                    log_message.append(f"{name}: {value}")
+                    self.log_metrics(
+                        mlflow_client,
+                        root_run_id,
+                        name,
+                        value,
+                    )
+                logger.info(", ".join(log_message))
+
+            log_message = [
+                f"End of training",
+            ]
+
+            # log metrics at the pipeline level
+            for name, value in train_metrics.get_global().items():
+                log_message.append(f"{name}: {value}")
+                self.log_metrics(
+                    mlflow_client,
+                    root_run_id,
+                    name,
+                    value,
+                    pipeline_level=True,
+                )
+
+            for name, value in test_metrics.get_global().items():
+                log_message.append(f"{name}: {value}")
+                self.log_metrics(
+                    mlflow_client,
+                    root_run_id,
+                    name,
+                    value,
+                    pipeline_level=True,
+                )
+            logger.info(", ".join(log_message))
+
+    def test(self):
+        """Test the trained model and report test loss and accuracy"""
+        test_metrics = RunningMetrics(
+            ["loss", "accuracy", "precision", "recall"], prefix="test"
+        )
+
+        self.model_bottom.eval()
```

Comment:
related to the concern about making it easy to understand, I would go as far as commenting here explain why we're doing both .eval() calls, compared to ccfraud vertical, you could add comments to both host and contributor scripts explaining some of the designs so that users wouldn't have to digest all the code to figure that out.